### DISCUSSION TITLE
How do I pass the content type of a file to Graph API in an uploadSession?

### DISCUSSION THREAD

**Discussion by Ghostbird** [microsoftgraph/msgraph-sdk-dotnet#2646] [Q&A] [answered]
Our system uploads files to OneDrive using the Graph API upload session. But I cannot find out how to set the correct content type. The `MimeType` of the resulting `DriveItem` is set to `application/octet-stream` instead of the correct image content type. Which makes sense, as I didn't tell Graph API what the content type is. I cannot find any documentation on how to pass the content type. I've seen that when doing "small file uploads" with lower-level HTTP, you can directly set the Content-Type HTTP Header. However for the large file upload API, all that is abstracted away.

How can I pass the content type correctly to Graph API when I upload the file using an `UploadSession`?

**Comment by andrueastman** [comment]
Thanks for raising this @Ghostbird 

Any chance you've come across the docs here? The should have info how to do large file uploads using the SDKs.

https://learn.microsoft.com/en-us/graph/sdks/large-file-upload?tabs=csharp

**Comment by Ghostbird** [comment]
Of course, however, there's nothing there about the content type.

**Comment by Ghostbird** [comment, marked as answer]
Whelp. I think Microsoft Graph is spectacularly unsuitable for storing files.

I'm not doing _anything_ special. I'm just uploading an image. But the content type returned is "application/octet-stream" which is _spectacularly_ useless.

The [File documentation](https://learn.microsoft.com/en-us/graph/api/resources/file?view=graph-rest-1.0) states

> The MIME type for the file. This is determined by logic on the server and might not be the value provided when the file was uploaded. Read-only.

Read-only is evident when trying to update it with the correct content type yields:

> Status Code: BadRequest
> Microsoft.Graph.ServiceException: Code: notSupported
> Message: Operation not supported

_WHY_ would _anyone_ design a file storage mechanism in such a way that it explicitly warps the file metadata into something incorrect when storing a file? This seems utterly mental to me!

This guarantees that if I ever have to choose a file storage for a project, it will NOT be through Microsoft Graph.

**Reply by Ghostbird** [reply]
Note: It depends on the actual file. It seems that Graph API likes to guess the content type based on the file name. That would be fine if no content type was provided. It would be doable if I could correct it myself later. But it's bananas that it's unchangeable.

**Reply by Ghostbird** [reply]
Even worse, it's unpredictable. Behaviour is different based on which specific `MicrosoftGraphDrive` I'm writing the files to. That SHOULD have been abstracted away correctly in the Graph API layer.

**Reply by ovdixon** [reply]
Did you ever find a solution to this, we are encountering the same issue?
